store/db.go

```diff
@@ -32,129 +28,38 @@ const (
-	// error when a lock is already held.
-	fl *lock.FileLock
-	// This lock is to avoid concurrent access from multiple goroutines.
-	sync.Mutex
-}
-
-func newDBLock(dirPath string) (*dbLock, error) {
-	l, err := lock.NewLock(dirPath, lock.Dir)
-	if err != nil {
-		return nil, err
-	}
-	return &dbLock{fl: l}, nil
-}
-
-func (dl *dbLock) lock() error {
-	dl.Lock()
-	if err := dl.fl.ExclusiveLock(); err != nil {
-		dl.Unlock()
-		return err
-	}
-	return nil
-}
-
-func (dl *dbLock) unlock() {
-	if err := dl.fl.Unlock(); err != nil {
-		// TODO(sgotti) what'll happen when df.fl.Unlock fails? From
-		// man 2 flock looks like it'll happen only when the underlying
-		// fd has been closed (in this case the lock has been released
-		// when the fd has been closed, assuming no dup fd due to
-		// forking etc...).
-
-		// If there're other cases where it fails without unlocking,
-		// there's no simple way to handle them.
-		// Possible solutions:
-		// * panic (done here)
-		// * try to close the lock (and related fd), panic if close
-		// fails and create a new lock.
-		//
-		// Passing a specific error to the caller and let it recover
-		// creating a new store instance is tricky because we
-		// don't know the lock state and cannot be sure on how to clean
-		// this store instance
-
-		panic(fmt.Errorf("failed to unlock the db flock: %v", err))
-	}
-	dl.Unlock()
-}
-
 type DB struct {
 	dbdir string
-	dl    *dbLock
-	sqldb *sql.DB
 }
 
 func NewDB(dbdir string) (*DB, error) {
 	if err := os.MkdirAll(dbdir, defaultPathPerm); err != nil {
 		return nil, err
 	}
-
-	dl, err := newDBLock(dbdir)
-	if err != nil {
-		return nil, err
-	}
-
-	return &DB{dbdir: dbdir, dl: dl}, nil
-}
-
-func (db *DB) Open() error {
-	if err := db.dl.lock(); err != nil {
-		return err
-	}
-
-	sqldb, err := sql.Open("ql", filepath.Join(db.dbdir, DbFilename))
-	if err != nil {
-		db.dl.unlock()
-		return err
-	}
-	db.sqldb = sqldb
-
-	return nil
-}
-
-func (db *DB) Close() error {
-	if db.sqldb == nil {
-		panic("cas db, Close called without an open sqldb")
-	}
-
-	if err := db.sqldb.Close(); err != nil {
-		return errwrap.Wrap(errors.New("cas db close failed"), err)
-	}
-	db.sqldb = nil
-
-	// Don't close the flock as it will be reused.
-	db.dl.unlock()
-	return nil
-}
-
-func (db *DB) Begin() (*sql.Tx, error) {
-	return db.sqldb.Begin()
+	return &DB{dbdir: dbdir}, nil
 }
 
 type txfunc func(*sql.Tx) error
 
 // Do Opens the db, executes DoTx and then Closes the DB
+// To support a multiprocess and multigoroutine model on a single file access
+// database like ql there's the need to exlusively lock, open, close, unlock the
+// db for every transaction. For this reason every db transaction should be
+// fast to not block other processes/goroutines.
 func (db *DB) Do(fns ...txfunc) error {
-	err := db.Open()
+	l, err := lock.ExclusiveLock(db.dbdir, lock.Dir)
 	if err != nil {
 		return err
 	}
-	defer db.Close()
+	defer l.Close()
```

Comment:
That's a problem everywhere in rkt code with pkg/lock since we always defer Close without caring/returning the error, and from the posix manual looks like the state will be undefined so close shouldn't be retried (needs to find a way to test if the lock is released or not).

For just error reporting, with defer I can use a named return value for error and do something like:

```
defer func() {
	if cerr := l.Close(); cerr != nil && err == nil {
		err = cerr
	}
}()
```

I'm just not sure it's worth or should be done in all other rkt places.



> Let's just hope this doesn't fail; at least the code is much prettier and I think/hope less error prone (since this'll primarily be called by short lived processes that close their fd's on exit anyways)

Yeah, with the execption of the api-service that opens a store instance at start and uses it for all its life.